File: python/jindosh.py
'''
    This solution to the Jindosh Riddle works in Python 2.7 or 3.x.
'''

import itertools

NAME = 'name'
COLOR = 'color'
DRINK = 'drink'
HEIRLOOM = 'heirloom'
ORIGIN = 'origin'
POSITION = 'position'

axes = {
    NAME: set(('winslow', 'marcolla', 'contee', 'natsiou', 'finch')),
    COLOR: set(('red', 'green', 'purple', 'blue', 'white')),
    DRINK: set(('whiskey', 'rum', 'beer', 'absinthe', 'wine')),
    HEIRLOOM: set(('diamond', 'tin', 'pendant', 'ring', 'medal')),
    ORIGIN: set(('dunwall', 'dabokva', 'fraeport', 'karnaca', 'baleton')),
    POSITION: set((1, 2, 3, 4, 5))
}

def get_axis(value):
    '''
    Given a value, return the axis it belongs to.
    '''

    for axis in axes:
        if value in axes[axis]:
            return axis
    raise ValueError(value)

class ConstraintViolationError(Exception):
    '''
    This exception is thrown by a Constraint whenever the riddle
    solution matrix it's applied against violates it.
    '''

    pass

class Constraint:
    '''
    A Constraint is an object with a single method, `apply()`, that
    takes a potential solution Matrix and applies a constraint against
    it.

    Constraints can optionally make alterations to proposed solutions;
    this represents "deductions" that the solver is making about the
    unknown properties of a proposed solution.

    The `apply()` method should raise a ConstraintViolationError if
    the proposed solution violates the constraint. Otherwise, it should
    return True if the solution was modified (mutated) based on the
    constraint, or False if not.
    '''

    def apply(self, matrix):
        raise NotImplementedError()

class ABConstraint(Constraint):
    '''
    A bi-directional constraint consisting of two properties and values.

    This is an abstract class.
    '''

    def __init__(self, a, b):
        self.a_axis = get_axis(a)
        self.b_axis = get_axis(b)
        self.a = a
        self.b = b

    def _apply(self, matrix, a_axis, a, b_axis, b):
        '''
        Apply the constraint with the given parameters on the given Matrix,
        returning True if the Matrix was mutated and False otherwise.
        '''

        raise NotImplementedError()

    def apply(self, matrix):
        a = self._apply(matrix, self.a_axis, self.a, self.b_axis, self.b)
        b = self._apply(matrix, self.b_axis, self.b, self.a_axis, self.a)
        return a or b

class SimpleConstraint(ABConstraint):
    '''
    A constraint stating that a person with property `a` along
    `a_axis` must also have property `b` along `b_axis`.
    '''

    @staticmethod
    def _apply(matrix, a_axis, a, b_axis, b):
        col = matrix.get_column(a, a_axis)
        if col:
            return col.set_or_verify(b_axis, b)
        return False

class NeighborConstraint(ABConstraint):
    '''
    A constraint stating that a person with property `a` along
    `a_axis` must have a neighbor with property `b` along `b_axis`.
    '''

    @classmethod
    def _apply_to_neighbor(cls, neighbor, other, axis, value):
        other_val = other[axis]
        if other_val is not None and other_val != value:
            return neighbor.set_or_verify(axis, value)

    @classmethod
    def _apply(cls, matrix, a_axis, a, b_axis, b):
        col = matrix.get_column(a, a_axis)
        if col:
            neighbors = col.get_neighbors()
            if len(neighbors) == 1:
                neighbor = neighbors[0]
                if neighbor[b_axis] is None:
                    return neighbor.set_or_verify(b_axis, b)
            elif len(neighbors) == 2:
                n1 = cls._apply_to_neighbor(neighbors[0], neighbors[1],
                                            b_axis, b)
                n2 = cls._apply_to_neighbor(neighbors[1], neighbors[0],
                                            b_axis, b)
                return n1 or n2
        return False

class PurpleIsLeftOfBlueConstraint(Constraint):
    '''
    A constraint stating that the person wearing purple must be
    immediately to the left of the person wearing blue.
    '''

    def apply(self, matrix):
        purple = matrix.get_column('purple', COLOR)
        blue = matrix.get_column('blue', COLOR)
        if purple and blue:
            if purple[POSITION] and blue[POSITION]:
                # The riddle is ambiguous about whether the purple
                # person must be *immediately* to the left of the 
                # blue person, or just "to the left of" in general,
                # but in order to obtain one unique result, we
                # need to interpret it as *immediately*.
                if purple[POSITION] != blue[POSITION] - 1:
                    raise ConstraintViolationError(
                        'purple must be left of blue'
                    )

constraints = [
    SimpleConstraint('contee', 'red'),
    SimpleConstraint(1, 'natsiou'),
    SimpleConstraint(2, 'green'),
    SimpleConstraint(3, 'beer'),
    SimpleConstraint('wine', 'purple'),
    SimpleConstraint('dabokva', 'white'),
    SimpleConstraint('winslow', 'diamond'),
    SimpleConstraint('baleton', 'ring'),
    SimpleConstraint('finch', 'absinthe'),
    SimpleConstraint('dunwall', 'whiskey'),
    SimpleConstraint('marcolla', 'fraeport'),
    NeighborConstraint('tin', 'dabokva'),
    NeighborConstraint('medal', 'karnaca'),
    NeighborConstraint('rum', 'karnaca'),
    PurpleIsLeftOfBlueConstraint(),
]

class MatrixColumn:
    '''
    A column of a Matrix, representing a particular person in a
    potential solution for the Jindosh riddle.
    '''

    def __init__(self, matrix, index):
        self.matrix = matrix
        self.index = index

    def get_neighbors(self):
        '''
        Return any MatrixColumns corresponding to the known neighbors of
        this one.
        '''

        pos = self.matrix.axes[POSITION][self.index]
        if pos is not None:
            left = self.matrix.get_column(pos - 1, POSITION)
            right = self.matrix.get_column(pos + 1, POSITION)
            return [n for n in [left, right] if n is not None]
        return []

    def set_or_verify(self, axis, value):
        '''
        Set the given property axis to the given value on the Matrix
        that this MatrixColumn comes from.

        Returns True if the Matrix was mutated, False otherwise.
        '''

        curr_value = self.matrix.axes[axis][self.index]
        if curr_value is None:
            if self.matrix.get_column(value, axis) is not None:
                raise ConstraintViolationError(
                    'cannot set %s to %s, as it is '
                    'already present elsewhere' % (
                        axis,
                        value
                    )
                )
            self.matrix.axes[axis][self.index] = value
            return True
        if curr_value != value:
            raise ConstraintViolationError(
                'cannot set %s to %s, as it is already %s' % (
                    axis,
                    value,
                    curr_value
                )
            )
        return False

    def __getitem__(self, axis):
        return self.matrix.axes[axis][self.index]

    def __setitem__(self, axis, value):
        self.set_or_verify(axis, value)

class Matrix:
    '''
    Represents one potential solution to the Jindosh Riddle. Each
    column represents a particular individual across all axes of
    their persona (name, drink, heirloom, etc).

    Each cell in the Matrix is `None` if its value is not yet known.
    '''

    def __init__(self, **kwargs):
        self.axes = {}
        for kwarg in kwargs:
            self.axes[kwarg] = list(kwargs[kwarg])
        for axis in axes:
            if axis not in self.axes:
                self.axes[axis] = [None] * len(axes[axis])
        self.apply_constraints()

    def get_column(self, value, axis=None):
        '''
        Return a MatrixColumn corresponding to the person with the
        given property, or None if no such person exists.
        '''

        if axis is None:
            axis = get_axis(value)
        if value in self.axes[axis]:
            return MatrixColumn(self, self.axes[axis].index(value))
        return None

    @staticmethod
    def _apply_permutation(row, unfilled_spots, permutation):
        row = row[:]
        for spot, i in zip(unfilled_spots, range(len(permutation))):
            row[spot] = permutation[i]
        return row

    def permute(self, axis):
        '''
        This generator yields Matrices with the unfilled cells along the
        given axis filled-in.

        For instance, if three of five heirlooms have been filled out, this
        will yield the two permutations of the Matrix with all the cells
        filled-in. 
        '''

        all_values = axes[axis]
        filled_values = set([
            value for value in self.axes[axis]
            if value is not None
        ])
        unfilled_spots = [
            i for i in range(len(self.axes[axis]))
            if self.axes[axis][i] is None
        ]
        values_to_permute = all_values.difference(filled_values)

        for permutation in itertools.permutations(values_to_permute):
            newaxes = self.axes.copy()
            newaxes[axis] = self._apply_permutation(
                self.axes[axis],
                unfilled_spots,
                permutation
            )
            try:
                matrix = self.__class__(**newaxes)
            except ConstraintViolationError:
                continue
            yield matrix

    def apply_constraints(self):
        '''
        Continuously apply all constraints to this particular solution
        until none of them change the solution.
        '''

        keep_going = True
        while keep_going:
            keep_going = False
            for constraint in constraints:
                if constraint.apply(self):
                    keep_going = True

    def final_sanity_check(self):
        '''
        If all the cells in the Matrix are filled out, this ensures that
        they cover all possible values on each axis, e.g. that no two
        people have the same drink/heirloom/et cetera.
        '''

        for axis in axes:
            if set(self.axes[axis]) != axes[axis]:
                raise AssertionError(
                    'solution %s is invalid: %s' % (axis, self.axes[axis])
                )

    def __str__(self):
        lines = []
        for axis in axes:
            line = ["%10s" % axis]
            for val in self.axes[axis]:
                if val is None:
                    val = '??'
                line.append("%10s" % str(val))
            lines.append(' '.join(line))
        return '\n'.join(lines)

def solve_riddle():
    '''
    Find the Matrix corresponding to the unique solution of the
    Jindosh Riddle and return it.
    '''

    solution = None
    starting_matrix = Matrix(name=list(axes[NAME]))

    for heirloomed_matrix in starting_matrix.permute(HEIRLOOM):
        for positioned_matrix in heirloomed_matrix.permute(POSITION):
            for colored_matrix in positioned_matrix.permute(COLOR):
                for drinked_matrix in colored_matrix.permute(DRINK):
                    for final_matrix in drinked_matrix.permute(ORIGIN):
                        if solution is not None:
                            raise AssertionError('multiple solutions found')
                        solution = final_matrix

    solution.final_sanity_check()
    return solution

if __name__ == '__main__':
    print(solve_riddle())

File: haskell/jindosh.hs
-- This solution to the Jindosh Riddle is probably horrible Haskell,
-- because I'm only 200 pages into "Haskell Programming from
-- first principles" (haskellbook.com).

import Data.List
import Data.Maybe
import Text.Printf

-- Here we define the various aspects of an individual's persona as
-- separate types. This may be unnecessary, but after reading
-- the book's chapters on types and typeclasses, I thought it'd
-- be nice to get more hands-on experience with types, and also
-- useful for the compiler to ensure that I hadn't made certain
-- kinds of mistakes.
data Name = Winslow | Marcolla | Contee | Natsiou | Finch
  deriving (Show, Eq, Enum)

data Color = Red | Green | Purple | Blue | White
  deriving (Show, Eq, Enum)

data Drink = Whiskey | Rum | Beer | Absinthe | Wine
  deriving (Show, Eq, Enum)

data Heirloom = Diamond | Tin | Pendant | Ring | Medal
  deriving (Show, Eq, Enum)

data Origin = Dunwall | Dabokva | Fraeport | Karnaca | Baleton
  deriving (Show, Eq, Enum)

data Position = FarLeft | SecondFromLeft | Center | SecondFromRight | FarRight
  deriving (Show, Eq, Enum, Ord)

-- A Person represents a person in the Jindosh riddle whose various
-- properties may still be unknown or have a concrete value.
data Person = Person { name :: Maybe Name
                     , color :: Maybe Color
                     , drink :: Maybe Drink
                     , heirloom :: Maybe Heirloom
                     , origin :: Maybe Origin
                     , position :: Maybe Position }
  deriving (Show, Eq)

nullPerson = Person { name=Nothing
                    , color=Nothing
                    , drink=Nothing
                    , heirloom=Nothing
                    , origin=Nothing
                    , position=Nothing }

-- Return a lost of all possible values of an Enum. Haskell's type
-- inference will figure out which Enum we want the values of.
allValues :: (Enum a) => [a]
allValues = (enumFrom (toEnum 0))

-- A Property represents metadata about a particular aspect of a
-- person.
--
-- It seems like Haskell has a concept called "Lenses" which might
-- do something similar, but its documentation was confusing so I
-- went this route instead.
data Property x = Property { get :: Person -> Maybe x
                           , set :: Person -> x -> Person
                           , values :: [x] }

nameProp = Property { get=name
                    , set=(\p x -> p { name=Just x })
                    , values=allValues }

colorProp = Property { get=color
                     , set=(\p x -> p { color=Just x })
                     , values=allValues}

drinkProp = Property { get=drink
                     , set=(\p x -> p { drink=Just x })
                     , values=allValues }

heirloomProp = Property { get=heirloom
                        , set=(\p x -> p { heirloom=Just x })
                        , values=allValues }

originProp = Property { get=origin
                      , set=(\p x -> p { origin=Just x })
                      , values=allValues }

positionProp = Property { get=position
                        , set=(\p x -> p { position=Just x })
                        , values=allValues }

-- A Constraint takes a potential solution and returns Nothing if the
-- solution violates the constraint. Alternatively, it will return the
-- potential solution, or an alteration of the potential solution based
-- on inferences made by the constraint logic.
type Constraint = [Person] -> Maybe [Person]

-- This returns a Constraint stating that a person with Property `aprop`
-- set to `a` must also have Property `bprop` set to `b`. It will
-- alter the proposed solution based on inferences made by this
-- constraint.
simpleConstraint :: (Eq x, Eq y) => Property x -> x -> Property y -> y -> Constraint
simpleConstraint aprop a bprop b =
  let
    isConstraintViolated person =
      let
        oneWay aprop a bprop b =
          if ((get aprop) person) == Just a &&
             ((get bprop) person) /= Nothing &&
             ((get bprop) person) /= Just b then True else False
      in
        if oneWay aprop a bprop b || oneWay bprop b aprop a
          then True else False
    applyConstraint person =
      let
        oneWay aprop a bprop b person =
          if ((get aprop) person) == Just a
            then ((set bprop) person b)
            else person
        in
          oneWay aprop a bprop b (oneWay bprop b aprop a person)
    constraint people =
      if any isConstraintViolated people
        then Nothing
        else Just (map applyConstraint people)
  in
    constraint

leftOf :: Position -> Position
leftOf position = pred position

rightOf :: Position -> Position
rightOf position = succ position

-- A SideOperator is just an infix operator that tells us whether
-- a person is to the left or right of a particular Position.
type SideOperator = Person -> Position -> Bool

isLeftOf :: SideOperator
isLeftOf person position =
  if position == FarLeft then False else
    (get positionProp) person == Just (leftOf position)

isRightOf :: SideOperator
isRightOf person position =
  if position == FarRight then False else
    (get positionProp) person == Just (rightOf position)

getNeighbors :: Person -> [Person] -> [Person]
getNeighbors person people =
  let
    getPosition x = (get positionProp) x
    isNeighbor position person =
      person `isLeftOf` position || person `isRightOf` position
  in
    case getPosition person of
      Nothing -> []
      Just x -> filter (isNeighbor x) people

-- This returns a Constraint stating that a person with Property `aprop`
-- set to `a` must have a neighbor whose Property `bprop` is set to `b`.
neighborConstraint :: (Eq x, Eq y) => Property x -> x -> Property y -> y -> Constraint
neighborConstraint aprop a bprop b =
  let
    isViolatedByNeighbor person neighbor =
      let
        oneWay aprop a bprop b =
          if ((get aprop) person) == Just a &&
             ((get bprop) neighbor) /= Nothing &&
             ((get bprop) neighbor) /= Just b then True else False
      in
        oneWay aprop a bprop b || oneWay bprop b aprop a
    isViolatedByEitherNeighbor person neighbors =
      let
        oneWay aprop a bprop b =
          let
            bs = catMaybes (map (get bprop) neighbors)
          in
            if ((get aprop) person) == Just a &&
               length bs == 2 &&
               (not (b `elem` bs)) then True else False
      in
        oneWay aprop a bprop b || oneWay bprop b aprop a
    isViolatedByNeighbors person neighbors =
      if length neighbors == 1 then
        isViolatedByNeighbor person (head neighbors)
      else
        isViolatedByEitherNeighbor person neighbors
    isViolatedByAnyone people =
      any (\x -> isViolatedByNeighbors x (getNeighbors x people)) people
    constraint people =
      if isViolatedByAnyone people then Nothing else
        -- TODO: Consider inferring values based on this constraint
        -- to speed things up.
        Just people
  in
    constraint

findPerson :: (Eq x) => Property x -> x -> [Person] -> Maybe Person
findPerson prop value people =
  if null people
    then Nothing
  else
    let
      person = head people
    in
      if ((get prop) person) == Just value
        then Just person
        else findPerson prop value (tail people)

-- This returns a Constraint stating that a person with Property `aprop`
-- set to `a` must be to the left or right of a person whose
-- Property `bprop` is set to `b`.
sideConstraint :: (Eq x, Eq y) => Property x -> x -> SideOperator -> Property y -> y -> Constraint
sideConstraint aprop a sideOp bprop b =
  let
    checkPosition personA personB defaultVal =
      let
        posA = (get positionProp) personA
        posB = (get positionProp) personB
      in
        if isNothing posA || isNothing posB then defaultVal else
          if personA `sideOp` (fromJust posB) then defaultVal else Nothing
    constraint people =
      let
        personA = findPerson aprop a people
        personB = findPerson bprop b people
      in
        if isNothing personA || isNothing personB then Just people else
          checkPosition (fromJust personA) (fromJust personB) (Just people)
  in
    constraint

-- Continuously apply Constraints to a potential solution until
-- a stable state is reached (that is, until none of the Constraints
-- return a changed solution, or until any Constraint is violated).
applyConstraints :: [Constraint] -> [Person] -> Maybe [Person]
applyConstraints constraints people =
  let
    applyConstraint constraint people =
      case people of
        Nothing -> Nothing
        Just x -> constraint x
    nextPeople =
      foldr applyConstraint (Just people) constraints
  in
    case nextPeople of
      Nothing -> Nothing
      Just x -> if x == people
        -- Nothing changed when we applied the constraints, so
        -- just return them as-is.
        then Just people
        -- Something changed when applying the constraints, so
        -- keep applying them.
        else applyConstraints constraints x

fillAbsentValues :: Property x -> [Person] -> [x] -> [Person]
fillAbsentValues prop people propValues =
  if null people then []
    else
      let
        person = head people
      in
        case ((get prop) person) of
          Nothing -> ((set prop) person (head propValues)):fillAbsentValues prop (tail people) (tail propValues)
          Just x -> person:fillAbsentValues prop (tail people) propValues

-- This returns a list of potential solutions with any unknown
-- values along a given Property filled-in.
--
-- For instance, if three of five heirlooms have been filled-in, this
-- will return a list containing the two permutations of the solution
-- with all heirlooms filled-in.
permuteProperty :: (Eq x) => Property x -> [Person] -> [[Person]]
permuteProperty prop people =
  let
    getValue = get prop
    allValues = values prop
    currentValues = map getValue people
    isValueAbsent value =
      not ((Just value) `elem` currentValues)
    valuesToPermute =
      filter isValueAbsent allValues
  in
    map (fillAbsentValues prop people) (permutations valuesToPermute)

solveForProperty :: (Eq x) => [Constraint] -> Property x -> [[Person]] -> [[Person]]
solveForProperty constraints prop candidates =
  let
    permutedCandidates = concat (map (permuteProperty prop) candidates)
  in
    catMaybes (map (applyConstraints constraints) permutedCandidates)

constraints :: [Constraint]
constraints = [ simpleConstraint nameProp Contee colorProp Red
              , simpleConstraint positionProp FarLeft nameProp Natsiou
              , simpleConstraint positionProp SecondFromLeft colorProp Green
              , simpleConstraint positionProp Center drinkProp Beer
              , simpleConstraint drinkProp Wine colorProp Purple
              , simpleConstraint originProp Dabokva colorProp White
              , simpleConstraint nameProp Winslow heirloomProp Diamond
              , simpleConstraint originProp Baleton heirloomProp Ring
              , simpleConstraint nameProp Finch drinkProp Absinthe
              , simpleConstraint originProp Dunwall drinkProp Whiskey
              , simpleConstraint nameProp Marcolla originProp Fraeport
              , neighborConstraint heirloomProp Tin originProp Dabokva
              , neighborConstraint heirloomProp Medal originProp Karnaca
              , neighborConstraint drinkProp Rum originProp Karnaca
              , sideConstraint colorProp Purple isLeftOf colorProp Blue
              ]

initialPeople :: [Person]
initialPeople =
  let
    placedPeople = map ((set positionProp) nullPerson) (values positionProp)
  in
    (fromJust (applyConstraints constraints placedPeople))

solns :: [[Person]]
solns =
  (solveForProperty constraints originProp
    (solveForProperty constraints drinkProp
      (solveForProperty constraints colorProp
        (solveForProperty constraints heirloomProp
          (solveForProperty constraints nameProp [initialPeople])
        )
      )
    )
  )

-- The following display code is super weird because I don't know what
-- Monads are yet, and things that I thought would work--like
-- issuing a `printf` in a function called by `map`--mysteriously
-- don't. Because side effects are bad, I guess? I have no idea.
display :: [Person] -> IO ()
display people =
  let
    getPropValue prop person =
      let
        value = ((get prop) person)
      in
        if value == Nothing then "??" else show (fromJust value)
    displayProp :: (Show x, Eq x) => Property x -> Person -> IO ()
    displayProp prop person =
      do
        printf "%20s " (getPropValue prop person)
    displayRow prop people =
      if null people then
        printf "\n"
        else
          do
            displayProp prop (head people)
            displayRow prop (tail people)
  in
    do
      displayRow nameProp people
      displayRow heirloomProp people
      displayRow drinkProp people
      displayRow originProp people
      displayRow colorProp people
      displayRow positionProp people

displayMany :: [[Person]] -> IO ()
displayMany candidates =
  if null candidates then return () else
    do
      display (head candidates)
      printf "\n"
      displayMany (tail candidates)

main =
  do
    if length solns /= 1 then
      do
        displayMany solns
        printf "WARNING: %d solutions found.\n" (length solns)
    else
      display (head solns)
